commit dbbbf5d5756fb31782c16ef969e5c9d4d04d3f9d
Author: Deepflex <hqdevnews@gmail.com>
Date:   Sat Aug 13 15:26:34 2016 +0700

    netd update
    
    Change-Id: I3c9b76c8a2644fe1335a85df177e1fb580d33049

diff --git a/server/RouteController.cpp b/server/RouteController.cpp
index 889779d..71e78e2 100644
--- a/server/RouteController.cpp
+++ b/server/RouteController.cpp
@@ -45,7 +45,7 @@ namespace {
 // BEGIN CONSTANTS --------------------------------------------------------------------------------
 
 const uint32_t RULE_PRIORITY_VPN_OVERRIDE_SYSTEM = 10000;
-const uint32_t RULE_PRIORITY_VPN_OVERRIDE_OIF    = 10500;
+//const uint32_t RULE_PRIORITY_VPN_OVERRIDE_OIF    = 10500;
 const uint32_t RULE_PRIORITY_VPN_OUTPUT_TO_LOCAL = 11000;
 const uint32_t RULE_PRIORITY_SECURE_VPN          = 12000;
 const uint32_t RULE_PRIORITY_EXPLICIT_NETWORK    = 13000;
@@ -556,6 +556,7 @@ WARN_UNUSED_RESULT int modifyOutputInterfaceRules(const char* interface, uint32_
     fwmark.permission = permission;
     mask.permission = permission;
 
+/*
     // If this rule does not specify a UID range, then also add a corresponding high-priority rule
     // for UID. This covers forwarded packets and system daemons such as the tethering DHCP server.
     if (uidStart == INVALID_UID && uidEnd == INVALID_UID) {
@@ -565,6 +566,7 @@ WARN_UNUSED_RESULT int modifyOutputInterfaceRules(const char* interface, uint32_
             return ret;
         }
     }
+*/
 
     return modifyIpRule(add ? RTM_NEWRULE : RTM_DELRULE, RULE_PRIORITY_OUTPUT_INTERFACE, table,
                         fwmark.intValue, mask.intValue, IIF_NONE, interface, uidStart, uidEnd);
diff --git a/server/SoftapController.cpp b/server/SoftapController.cpp
index b05118e..08e4942 100755
--- a/server/SoftapController.cpp
+++ b/server/SoftapController.cpp
@@ -42,6 +42,7 @@
 #include "wifi.h"
 #include "wifi_fst.h"
 #include "ResponseCode.h"
+#include "cutils/properties.h"
 
 #include "SoftapController.h"
 
@@ -240,6 +241,8 @@ bool SoftapController::isSoftapStarted() {
     return (mPid != 0);
 }
 
+#define USE_MTK_WLAN   1
+
 /*
  * Arguments:
  *  argv[2] - wlan interface
@@ -253,6 +256,12 @@ int SoftapController::setSoftap(int argc, char *argv[]) {
     int hidden = 0;
     int channel = AP_CHANNEL_DEFAULT;
 
+#ifdef USE_MTK_WLAN
+    char ap_channel[PROPERTY_VALUE_MAX] = {'\0'};
+    const char AP_CHANNEL_PROP_NAME[]="wifi.tethering.channel";
+    char mode;
+#endif
+
     if (argc < 5) {
         ALOGE("Softap set is missing arguments. Please use:");
         ALOGE("softap <wlan iface> <SSID> <hidden/broadcast> <channel> <wpa2?-psk|open> <passphrase>");
@@ -263,11 +272,31 @@ int SoftapController::setSoftap(int argc, char *argv[]) {
         hidden = 1;
 
     if (argc >= 5) {
+    #ifndef USE_MTK_WLAN
         channel = atoi(argv[5]);
         if (channel <= 0)
             channel = AP_CHANNEL_DEFAULT;
+    #else
+        if (property_get(AP_CHANNEL_PROP_NAME, ap_channel, NULL)){
+            channel = atoi(ap_channel);
+        } else {
+            channel = atoi(argv[5]);
+        }
+
+        if (channel < 0) {
+            channel = AP_CHANNEL_DEFAULT;
+            mode = 'g';
+        } else if (channel >= 0 && channel <= 14) {
+            /* 2.4G channel, 0 will trigger MTK auto channel selection */
+            mode = 'g';
+        } else {
+            /* 5G channel */
+            mode = 'a';
+        }
+    #endif
     }
 
+#ifndef USE_MTK_WLAN
     std::string wbuf(StringPrintf("interface=%s\n"
             "driver=nl80211\n"
             "ctrl_interface=/data/misc/wifi/hostapd\n"
@@ -278,6 +307,37 @@ int SoftapController::setSoftap(int argc, char *argv[]) {
             "ignore_broadcast_ssid=%d\n"
             "wowlan_triggers=any\n",
             argv[2], argv[3], channel, (channel <= 14) ? 'g' : 'a', hidden));
+#else
+/*
+ * Arguments:
+ *  argv[2] - wlan interface
+ *  argv[3] - SSID
+ *  argv[4] - Broadcast/Hidden
+ *  argv[5] - Channel
+ *  argv[6] - Security
+ *  argv[7] - Key
+ *  argv[8] - Channel width
+ *  argv[9] - Max connection
+ */
+    std::string wbuf;
+    if (argc > 9) {
+        wbuf = StringPrintf("interface=%s\ndriver=nl80211\nctrl_interface="
+            "/data/misc/wifi/hostapd\nssid=%s\nchannel=%d\nieee80211n=1\n"
+            "hw_mode=%c\nignore_broadcast_ssid=%d\nmax_num_sta=%d\n"
+            "eap_server=1\nwps_state=2\nconfig_methods=display physical_display push_button\n"
+            "device_name=AndroidAP\nmanufacturer=MediaTek Inc.\nmodel_name=MTK Wireless Model\n"
+            "model_number=66xx\nserial_number=1.0\ndevice_type=10-0050F204-5\nwowlan_triggers=disconnect\n",
+            "ap0", argv[3], channel, mode, hidden, atoi(argv[9]));
+    } else {
+        wbuf = StringPrintf("interface=%s\ndriver=nl80211\nctrl_interface="
+            "/data/misc/wifi/hostapd\nssid=%s\nchannel=%d\nieee80211n=1\n"
+            "hw_mode=%c\nignore_broadcast_ssid=%d\n"
+            "eap_server=1\nwps_state=2\nconfig_methods=display physical_display push_button\n"
+            "device_name=AndroidAP\nmanufacturer=MediaTek Inc.\nmodel_name=MTK Wireless Model\n"
+            "model_number=66xx\nserial_number=1.0\ndevice_type=10-0050F204-5\nwowlan_triggers=disconnect\n",
+            "ap0", argv[3], channel, mode, hidden);
+    }
+#endif
 
     std::string fbuf;
     if (argc > 7) {
